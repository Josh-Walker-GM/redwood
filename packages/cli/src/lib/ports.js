import portfinder from 'portfinder'

/**
 * Finds a free port
 * @param  {[number]}   requestedPort Port to start searching from
 * @param  {[number[]]} excludePorts  Array of port numbers to exclude
 * @return {[number]}                 A free port equal or higher than requestedPort but not within excludePorts. If no port can be found then returns -1
 */
export async function getFreePort(requestedPort, excludePorts = []) {
  try {
    let freePort = await portfinder.getPortPromise({
      port: requestedPort,
    })
    if (excludePorts.includes(freePort)) {
      freePort = await getFreePort(freePort + 1, excludePorts)
    }
    return freePort
  } catch (error) {
    return -1
  }
}

/**
 * Determines if a port is available or in use
 * @param {number} port The port number to check
 * @return {boolean} True if the port is available, false otherwise
 */
export async function isPortFree(port) {
  return (await portfinder.getPortPromise({ port })) === port
}
