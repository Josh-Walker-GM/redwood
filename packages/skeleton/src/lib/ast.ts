import fs from 'fs'

import { parse } from '@babel/parser'
import traverse from '@babel/traverse'
import {
  assertJSXOpeningElement,
  assertJSXIdentifier,
  assertIdentifier,
  isJSXAttribute,
  isJSXSpreadAttribute,
  JSXElement,
  ExportNamedDeclaration,
  isVariableDeclaration,
  isVariableDeclarator,
  isIdentifier,
  File,
  VariableDeclarator,
  ExportDefaultDeclaration,
} from '@babel/types'

export function getASTFromCode(code: string) {
  return parse(code, {
    sourceType: 'unambiguous',
    // TODO: Check these plugin options are optimal, I doubt they are
    plugins: [
      'jsx',
      'typescript',
      'nullishCoalescingOperator',
      'objectRestSpread',
    ],
  })
}

export function getASTFromFile(filePath: string) {
  const code = fs.readFileSync(filePath, { encoding: 'utf8', flag: 'r' })
  return getASTFromCode(code)
}

export function getSpecificNamedExportDeclarations(ast: File, names: string[]) {
  const desiredNamedExports: Map<string, VariableDeclarator> = new Map()
  const allNamedExports: ExportNamedDeclaration[] = []
  traverse(ast, {
    ExportNamedDeclaration: (path) => {
      allNamedExports.push(path.node)
    },
  })
  allNamedExports.forEach((namedExport) => {
    if (
      namedExport.declaration != null &&
      isVariableDeclaration(namedExport.declaration) &&
      isVariableDeclarator(namedExport.declaration.declarations[0]) &&
      isIdentifier(namedExport.declaration.declarations[0].id) &&
      names.includes(namedExport.declaration.declarations[0].id.name)
    ) {
      desiredNamedExports.set(
        namedExport.declaration.declarations[0].id.name,
        namedExport.declaration.declarations[0]
      )
    }
  })
  return desiredNamedExports
}

export function getExportDefaultDeclaration(ast: File) {
  let defaultExport: ExportDefaultDeclaration | undefined
  traverse(ast, {
    ExportDefaultDeclaration: (path) => {
      defaultExport = path.node
    },
  })
  return defaultExport
}

export function getVariableDeclarator(ast: File, name: string) {
  let variableDeclarator: VariableDeclarator | undefined
  traverse(ast, {
    VariableDeclarator: (path) => {
      if (isIdentifier(path.node.id) && path.node.id.name === name) {
        variableDeclarator = path.node
      }
    },
  })
  return variableDeclarator
}

export function getJSXElementAttributes(
  element: JSXElement
): Map<string, string> {
  const attributes: Map<string, string> = new Map()
  assertJSXOpeningElement(element.openingElement)
  if (element.openingElement.attributes.length > 0) {
    element.openingElement.attributes.forEach((attribute) => {
      if (isJSXAttribute(attribute)) {
        assertJSXIdentifier(attribute.name)
        const key = attribute.name.name
        let value: string
        switch (attribute.value?.type) {
          case 'StringLiteral':
            value = attribute.value.value
            break
          case 'JSXExpressionContainer':
            assertIdentifier(attribute.value.expression)
            value = attribute.value.expression.name
            break
          case undefined: // Accept this because it could be a attribute with no value, i.e. notfound
            value = 'true'
            break
          default:
            throw new Error(
              `Unknown value type found while trying to process the ${key} property when extracting all attributes`
            )
        }
        attributes.set(key, value)
      } else if (isJSXSpreadAttribute(attribute)) {
        throw new Error(
          'JSXSpreadAttribute type is not yet supported when extracting all attributes'
        )
      } else {
        throw new Error(
          'Uknown attribute type found while trying to extract all attributes'
        )
      }
    })
  }
  return attributes
}
