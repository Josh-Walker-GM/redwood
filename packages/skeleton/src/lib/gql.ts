import { parse } from 'graphql'

export function parseGraphQL(gql: string) {
  return parse(gql)
}

export function getGraphQLQueryName(gql: string): string | undefined {
  const ast = parse(gql)
  for (const def of ast.definitions) {
    if (def.kind === 'OperationDefinition' && def.operation === 'query') {
      return def.name?.value
    }
  }
  return undefined
}
