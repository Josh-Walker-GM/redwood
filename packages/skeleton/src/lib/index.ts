import fs from 'fs'

import { parse } from '@babel/parser'

export function getFileAST(filePath: string) {
  const code = fs.readFileSync(filePath, { encoding: 'utf8', flag: 'r' })
  return parse(code, {
    sourceType: 'unambiguous',
    plugins: ['jsx', 'typescript'],
  })
}
