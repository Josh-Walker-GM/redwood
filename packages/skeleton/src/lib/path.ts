import path from 'path'

import { getBaseDir, getConfigPath } from '@redwoodjs/internal/dist/index'

/**
 * Determines the root path of a redwood project which contains the provided path
 * @param pathWithinProject A path within a redwood project
 * @returns The root path of the associated project
 */
export function getRootPath(pathWithinProject: string) {
  return pathWithinProject
    ? path.dirname(getConfigPath(pathWithinProject))
    : getBaseDir()
}
