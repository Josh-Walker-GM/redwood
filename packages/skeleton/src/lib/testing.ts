import { ensurePosixPath } from '@redwoodjs/internal/dist/paths'

export function stripAndFormatPathForTesting(
  filepath: string,
  projectFilepath: string
) {
  if (filepath.startsWith(projectFilepath)) {
    filepath = ensurePosixPath(filepath.substring(projectFilepath.length))
  }
  return filepath
}
