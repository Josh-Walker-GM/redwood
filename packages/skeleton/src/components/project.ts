import fs from 'fs'
import path from 'path'

import chalk from 'chalk'

import { getPaths } from '@redwoodjs/internal/dist/index'

import { getRootPath } from '../lib/path'

import { RedwoodSkeleton } from './base'
import { extractCells, RedwoodCell } from './cell'
import { extractFunctions, RedwoodFunction } from './function'
import { extractLayouts, RedwoodLayout } from './layout'
import { extractPages, RedwoodPage } from './page'
import { extractRouters, RedwoodRouter } from './router'
import { extractSides, RedwoodSide } from './side'

/**
 * Used to enumerate either JS or TS project types
 */
export enum RedwoodProjectType {
  TYPESCRIPT = 'typescript',
  JAVASCRIPT = 'javascript',
}

export class RedwoodProject extends RedwoodSkeleton {
  warnings: string[] = []
  errors: string[] = []

  readonly type: RedwoodProjectType

  private sides?: RedwoodSide[] | undefined
  private cells?: RedwoodCell[] | undefined
  private routers?: RedwoodRouter[] | undefined
  private layouts?: RedwoodLayout[] | undefined
  private pages?: RedwoodPage[] | undefined
  private functions?: RedwoodFunction[] | undefined

  public static getProject({
    pathWithinProject = '',
    full = false,
    readFromCache = true,
    insertIntoCache = true,
  }: {
    pathWithinProject?: string
    full?: boolean
    readFromCache?: boolean
    insertIntoCache?: boolean
  } = {}) {
    const projectCache = RedwoodProjectsCache.getInstance().projects
    const rootPath = getRootPath(pathWithinProject)

    if (readFromCache) {
      const cachedProject = projectCache.get(rootPath)
      if (cachedProject) {
        return cachedProject
      }
    }

    const project = new RedwoodProject(rootPath, full)
    if (insertIntoCache) {
      projectCache.set(project.filepath, project)
    }
    return project
  }

  private constructor(rootPath: string, full = false) {
    super(rootPath)

    // A project is typescript if we detect a tsconfig.json
    this.type =
      fs.existsSync(
        path.join(getPaths(this.filepath).web.base, 'tsconfig.json')
      ) ||
      fs.existsSync(
        path.join(getPaths(this.filepath).api.base, 'tsconfig.json')
      )
        ? RedwoodProjectType.TYPESCRIPT
        : RedwoodProjectType.JAVASCRIPT

    if (full) {
      this.sides = extractSides(this)
      this.cells = extractCells(this)
      this.layouts = extractLayouts(this)
      this.pages = extractPages(this)
      this.functions = extractFunctions(this)
      this.routers = extractRouters(this)
    }
  }

  getSides(forceExtract = false): RedwoodSide[] {
    if (forceExtract) {
      return extractSides(this)
    }
    return this.sides === undefined ? extractSides(this) : this.sides
  }

  getCells(forceExtract = false): RedwoodCell[] {
    if (forceExtract) {
      return extractCells(this)
    }
    return this.cells === undefined ? extractCells(this) : this.cells
  }

  getRouters(forceExtract = false): RedwoodRouter[] {
    if (forceExtract) {
      return extractRouters(this)
    }
    return this.routers === undefined ? extractRouters(this) : this.routers
  }

  getLayouts(forceExtract = false): RedwoodLayout[] {
    if (forceExtract) {
      return extractLayouts(this)
    }
    return this.layouts === undefined ? extractLayouts(this) : this.layouts
  }

  getPages(forceExtract = false): RedwoodPage[] {
    if (forceExtract) {
      return extractPages(this)
    }
    return this.pages === undefined ? extractPages(this) : this.pages
  }

  getFunctions(forceExtract = false): RedwoodFunction[] {
    if (forceExtract) {
      return extractFunctions(this)
    }
    return this.functions === undefined
      ? extractFunctions(this)
      : this.functions
  }

  // Diagnostics

  getInformation(): string {
    return '' // TODO: Implement, gets the info of all other components too
  }

  printInformation(): void {
    console.log(
      `${chalk.bgCyan('[Info]')}\t${this.name}\t${chalk.dim(this.filepath)}`
    )
  }

  printWarnings(cascade = false): void {
    if (this.warnings.length > 0) {
      const titleLine = `${chalk.bgYellow('[Warn]')}\t${this.name} ${chalk.dim(
        this.filepath
      )}`
      const warningLines = this.warnings.map((warning, index) => {
        return ` (${index + 1}) ${warning}\n`
      })
      console.log(titleLine.concat('\n', ...warningLines).trimEnd())
    }
    if (cascade) {
      this.cells?.forEach((cell) => {
        cell.printWarnings()
      })
      this.functions?.forEach((func) => {
        func.printWarnings()
      })
      this.layouts?.forEach((layout) => {
        layout.printWarnings()
      })
      this.pages?.forEach((page) => {
        page.printWarnings()
      })
      this.routers?.forEach((router) => {
        router.printWarnings()
        router.routes.forEach((route) => {
          route.printWarnings()
        })
      })
      this.sides?.forEach((side) => {
        side.printWarnings()
      })
    }
  }

  printErrors(cascade = false): void {
    if (this.errors.length > 0) {
      const titleLine = `${chalk.bgRed('[Error]')}\t${this.name} ${chalk.dim(
        this.filepath
      )}`
      const errorLines = this.errors.map((error, index) => {
        return ` (${index + 1}) ${error}\n`
      })
      console.log(titleLine.concat('\n', ...errorLines).trimEnd())
    }
    if (cascade) {
      this.cells?.forEach((cell) => {
        cell.printErrors()
      })
      this.functions?.forEach((func) => {
        func.printErrors()
      })
      this.layouts?.forEach((layout) => {
        layout.printErrors()
      })
      this.pages?.forEach((page) => {
        page.printErrors()
      })
      this.routers?.forEach((router) => {
        router.printErrors()
        router.routes.forEach((route) => {
          route.printErrors()
        })
      })
      this.sides?.forEach((side) => {
        side.printErrors()
      })
    }
  }
}

/**
 * Singelton class which contains a cache of {@link RedwoodProject} instances.
 *
 * This class is not exported and should only be available from static methods within {@link RedwoodProject}, in order to keep the surface of the skeleton API cleaner and leaner.
 */
class RedwoodProjectsCache {
  private static instance: RedwoodProjectsCache

  public projects: Map<string, RedwoodProject>

  private constructor() {
    this.projects = new Map()
  }

  public static getInstance(): RedwoodProjectsCache {
    if (!RedwoodProjectsCache.instance) {
      RedwoodProjectsCache.instance = new RedwoodProjectsCache()
    }
    return RedwoodProjectsCache.instance
  }
}
