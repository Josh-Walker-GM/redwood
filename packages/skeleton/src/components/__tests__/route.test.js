import path from 'path'

import { stripAndFormatPathForTesting } from '../../lib/testing'
import { RedwoodProject } from '../project'

const FIXTURE_PATH = path.join(
  __dirname,
  '..',
  '..',
  '..',
  '..',
  '..',
  '__fixtures__'
)

describe.each([
  'empty-project',
  'example-todo-main',
  'example-todo-main-with-errors',
  'test-project',
])('From within the %s fixture', (PROJECT_NAME) => {
  const PROJECT_PATH = path.join(FIXTURE_PATH, PROJECT_NAME)

  const RWJS_CWD = process.env.RWJS_CWD
  beforeAll(() => {
    process.env.RWJS_CWD = PROJECT_PATH
  })
  afterAll(() => {
    process.env.RWJS_CWD = RWJS_CWD
  })

  const project = RedwoodProject.getProject({
    pathWithinProject: PROJECT_PATH,
    insertIntoCache: false,
    readFromCache: false,
  })
  const routers = project
    .getRouters()
    .sort((a, b) =>
      a.filepath > b.filepath ? 1 : b.filepath > a.filepath ? -1 : 0
    )

  describe.each(routers)('router $#', (router) => {
    const routes = router.routes
    routes.forEach((route) => {
      route.executeAdditionalChecks()
    })
    test('routes return the correct pages', () => {
      const routePageMap = new Map()
      routes.forEach((route) => {
        routePageMap.set(route, route.getPage())
      })
      routePageMap.forEach((value, key) => {
        key.filepath = stripAndFormatPathForTesting(key.filepath, PROJECT_PATH)
        if (value !== undefined) {
          value.filepath = stripAndFormatPathForTesting(
            value.filepath,
            PROJECT_PATH
          )
        }
      })
      expect(routePageMap).toMatchSnapshot()
    })
  })
})
