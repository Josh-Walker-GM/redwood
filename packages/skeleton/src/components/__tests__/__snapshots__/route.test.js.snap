// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From within the empty-project fixture router 0 routes return the correct pages 1`] = `
Map {
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": false,
    "isNotFound": true,
    "isPrivate": false,
    "name": "notfound",
    "pageIdentifier": "NotFoundPage",
    "path": undefined,
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/NotFoundPage/NotFoundPage.tsx",
    "name": "NotFoundPage",
    "warnings": [],
  },
}
`;

exports[`From within the example-todo-main fixture router 0 routes return the correct pages 1`] = `
Map {
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.js",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "home",
    "pageIdentifier": "HomePage",
    "path": "/",
    "prerender": true,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/HomePage/HomePage.tsx",
    "name": "HomePage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.js",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "typescriptPage",
    "pageIdentifier": "TypeScriptPage",
    "path": "/typescript",
    "prerender": true,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/TypeScriptPage/TypeScriptPage.tsx",
    "name": "TypeScriptPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [
      "Could not find page EditUserPage",
    ],
    "filepath": "/web/src/Routes.js",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "someOtherPage",
    "pageIdentifier": "EditUserPage",
    "path": "/somewhereElse",
    "prerender": true,
    "warnings": [],
  } => undefined,
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.js",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "fooPage",
    "pageIdentifier": "FooPage",
    "path": "/foo",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/FooPage/FooPage.tsx",
    "name": "FooPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.js",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "barPage",
    "pageIdentifier": "BarPage",
    "path": "/bar",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/BarPage/BarPage.tsx",
    "name": "BarPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.js",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": true,
    "name": "privatePage",
    "pageIdentifier": "PrivatePage",
    "path": "/private-page",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/PrivatePage/PrivatePage.tsx",
    "name": "PrivatePage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.js",
    "hasParameters": false,
    "isNotFound": true,
    "isPrivate": false,
    "name": "notfound",
    "pageIdentifier": "NotFoundPage",
    "path": undefined,
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/NotFoundPage/NotFoundPage.js",
    "name": "NotFoundPage",
    "warnings": [],
  },
}
`;

exports[`From within the example-todo-main-with-errors fixture router 0 routes return the correct pages 1`] = `
Map {
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.js",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "home",
    "pageIdentifier": "HomePage",
    "path": "/",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/HomePage/HomePage.js",
    "name": "HomePage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.js",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "home",
    "pageIdentifier": "HomePage",
    "path": "/",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/HomePage/HomePage.js",
    "name": "HomePage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [
      "Could not find page DoesNotExist",
    ],
    "filepath": "/web/src/Routes.js",
    "hasParameters": true,
    "isNotFound": false,
    "isPrivate": false,
    "name": "something-else",
    "pageIdentifier": "DoesNotExist",
    "path": "/{foo}/{foo}",
    "prerender": false,
    "warnings": [],
  } => undefined,
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.js",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "Routes",
    "pageIdentifier": "NotFoundPage",
    "path": undefined,
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/NotFoundPage/NotFoundPage.js",
    "name": "NotFoundPage",
    "warnings": [],
  },
}
`;

exports[`From within the test-project fixture router 0 routes return the correct pages 1`] = `
Map {
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "login",
    "pageIdentifier": "LoginPage",
    "path": "/login",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/LoginPage/LoginPage.tsx",
    "name": "LoginPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "signup",
    "pageIdentifier": "SignupPage",
    "path": "/signup",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/SignupPage/SignupPage.tsx",
    "name": "SignupPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "forgotPassword",
    "pageIdentifier": "ForgotPasswordPage",
    "path": "/forgot-password",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/ForgotPasswordPage/ForgotPasswordPage.tsx",
    "name": "ForgotPasswordPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "resetPassword",
    "pageIdentifier": "ResetPasswordPage",
    "path": "/reset-password",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/ResetPasswordPage/ResetPasswordPage.tsx",
    "name": "ResetPasswordPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "newContact",
    "pageIdentifier": "ContactNewContactPage",
    "path": "/contacts/new",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/Contact/NewContactPage/NewContactPage.tsx",
    "name": "ContactNewContactPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": true,
    "isNotFound": false,
    "isPrivate": false,
    "name": "editContact",
    "pageIdentifier": "ContactEditContactPage",
    "path": "/contacts/{id:Int}/edit",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/Contact/EditContactPage/EditContactPage.tsx",
    "name": "ContactEditContactPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": true,
    "isNotFound": false,
    "isPrivate": false,
    "name": "contact",
    "pageIdentifier": "ContactContactPage",
    "path": "/contacts/{id:Int}",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/Contact/ContactPage/ContactPage.tsx",
    "name": "ContactContactPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "contacts",
    "pageIdentifier": "ContactContactsPage",
    "path": "/contacts",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/Contact/ContactsPage/ContactsPage.tsx",
    "name": "ContactContactsPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "newPost",
    "pageIdentifier": "PostNewPostPage",
    "path": "/posts/new",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/Post/NewPostPage/NewPostPage.tsx",
    "name": "PostNewPostPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": true,
    "isNotFound": false,
    "isPrivate": false,
    "name": "editPost",
    "pageIdentifier": "PostEditPostPage",
    "path": "/posts/{id:Int}/edit",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/Post/EditPostPage/EditPostPage.tsx",
    "name": "PostEditPostPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": true,
    "isNotFound": false,
    "isPrivate": false,
    "name": "post",
    "pageIdentifier": "PostPostPage",
    "path": "/posts/{id:Int}",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/Post/PostPage/PostPage.tsx",
    "name": "PostPostPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "posts",
    "pageIdentifier": "PostPostsPage",
    "path": "/posts",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/Post/PostsPage/PostsPage.tsx",
    "name": "PostPostsPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": true,
    "isNotFound": false,
    "isPrivate": false,
    "name": "waterfall",
    "pageIdentifier": "WaterfallPage",
    "path": "/waterfall/{id:Int}",
    "prerender": true,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/WaterfallPage/WaterfallPage.tsx",
    "name": "WaterfallPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": true,
    "name": "profile",
    "pageIdentifier": "ProfilePage",
    "path": "/profile",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/ProfilePage/ProfilePage.tsx",
    "name": "ProfilePage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": true,
    "isNotFound": false,
    "isPrivate": false,
    "name": "blogPost",
    "pageIdentifier": "BlogPostPage",
    "path": "/blog-post/{id:Int}",
    "prerender": true,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/BlogPostPage/BlogPostPage.tsx",
    "name": "BlogPostPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "contactUs",
    "pageIdentifier": "ContactUsPage",
    "path": "/contact",
    "prerender": false,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/ContactUsPage/ContactUsPage.tsx",
    "name": "ContactUsPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "about",
    "pageIdentifier": "AboutPage",
    "path": "/about",
    "prerender": true,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/AboutPage/AboutPage.tsx",
    "name": "AboutPage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": false,
    "isNotFound": false,
    "isPrivate": false,
    "name": "home",
    "pageIdentifier": "HomePage",
    "path": "/",
    "prerender": true,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/HomePage/HomePage.tsx",
    "name": "HomePage",
    "warnings": [],
  },
  RedwoodRoute {
    "errors": [],
    "filepath": "/web/src/Routes.tsx",
    "hasParameters": false,
    "isNotFound": true,
    "isPrivate": false,
    "name": "notfound",
    "pageIdentifier": "NotFoundPage",
    "path": undefined,
    "prerender": true,
    "warnings": [],
  } => RedwoodPage {
    "errors": [],
    "filepath": "/web/src/pages/NotFoundPage/NotFoundPage.tsx",
    "name": "NotFoundPage",
    "warnings": [],
  },
}
`;
